policy-controller/src/main.rs

```diff
@@ -232,3 +231,130 @@ impl std::str::FromStr for LogFormat {
+    // Load and return a single private key.
+    let keys = rustls_pemfile::rsa_private_keys(&mut reader)?;
+    if keys.len() != 1 {
+        return Err(anyhow!("expected a single private key"));
+    }
+
+    Ok(rustls::PrivateKey(keys[0].clone()))
+}
+
+/// Bind the specified address and serve the admission controller using the
+/// provided kubernetes client.
+async fn bind_admission_controller(
+    bind_addr: SocketAddr,
+    client: kube::Client,
+) -> Result<(
+    SocketAddr,
+    impl std::future::Future<Output = ()> + Send + 'static,
+)> {
+    let tcp = TcpListener::bind(&bind_addr).await?;
+    let listen_addr = tcp.local_addr()?;
+    // Connection accept loop
+    let serve = async move {
+        loop {
+            let socket = match tcp.accept().await {
+                Ok((socket, _)) => socket,
+                Err(err) => {
+                    error!(%err, "Failed to accept connection");
+                    continue;
+                }
+            };
+            let client_addr = match socket.peer_addr() {
+                Ok(addr) => addr,
+                Err(error) => {
+                    error!(%error, "Failed to get peer address");
+                    continue;
+                }
+            };
+
+            tokio::spawn(
+                serve_admission_controller_conn(socket, client.clone())
+                    .instrument(info_span!("connection", client.addr = %client_addr)),
+            );
+        }
+    };
+
+    Ok((listen_addr, serve))
+}
+
+/// Serve an HTTP server for the admission controller on the given TCP
+/// connection.
+async fn serve_admission_controller_conn(socket: TcpStream, client: kube::Client) {
+    let tls = {
+        // Load public certificate.
+        let certs = match load_certs("/var/run/linkerd/tls/tls.crt")
+            .await
+            .with_context(|| "failed to load certificate")
+        {
+            Ok(certs) => certs,
+            Err(error) => {
+                error!(%error);
+                return;
+            }
+        };
+
+        // Load private key.
+        let key = match load_private_key("/var/run/linkerd/tls/tls.key")
+            .await
+            .with_context(|| "failed to load private key")
+        {
+            Ok(key) => key,
+            Err(error) => {
+                error!(%error);
+                return;
+            }
+        };
+
+        let mut cfg = match rustls::ServerConfig::builder()
+            .with_safe_defaults()
+            // Do not use client certificate authentication.
+            .with_no_client_auth()
+            .with_single_cert(certs, key)
+        {
+            Ok(cfg) => cfg,
+            Err(error) => {
+                error!(%error, "Failed to configure TLS");
+                return;
+            }
+        };
+        // Configure ALPN to accept HTTP/2, HTTP/1.1 in that order.
+        cfg.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec()];
+        TlsAcceptor::from(Arc::new(cfg))
+    };
```

Comment:
take it or leave it: we could probably reduce the amount of error handling boilerplate if we moved all of this into a function, like:
```suggestion
async fn serve_admission_controller_conn(socket: TcpStream, client: kube::Client) {
   async fn tls_acceptor() -> anyhow::Result<TlsAcceptor> {
        // Load public certificate.
        let certs = match load_certs("/var/run/linkerd/tls/tls.crt")
            .await
            .with_context(|| "failed to load certificate")?;

        // Load private key.
        let key = match load_private_key("/var/run/linkerd/tls/tls.key")
            .await
            .with_context(|| "failed to load private key")?;

        let mut cfg = match rustls::ServerConfig::builder()
            .with_safe_defaults()
            // Do not use client certificate authentication.
            .with_no_client_auth()
            .with_single_cert(certs, key)
            .with_context(|| "failed to configure TLS")?;

        // Configure ALPN to accept HTTP/2, HTTP/1.1 in that order.
        cfg.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec()];
        Ok(TlsAcceptor::from(Arc::new(cfg)))
    }

    let tls = match tls_acceptor().await {
        Ok(tls) => tls,
        Err(error) => {
            error!(%error);
            return;
        }
    };
```

Reply:
we could also just make the whole function return a `Result` and also be able to use `?` when actually accepting the connection, so that we can just have one `match` that actually prints the error...